hbase-client/src/main/java/org/apache/hadoop/hbase/client/backoff/HBaseServerExceptionPauseManager.java

```diff
@@ -0,0 +1,95 @@
+
+import static org.apache.hadoop.hbase.client.ConnectionUtils.SLEEP_DELTA_NS;
+import static org.apache.hadoop.hbase.client.ConnectionUtils.getPauseTime;
+
+import java.util.OptionalLong;
+import java.util.concurrent.TimeUnit;
+import org.apache.hadoop.hbase.HBaseServerException;
+import org.apache.hadoop.hbase.quotas.RpcThrottlingException;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@InterfaceAudience.Private
+public class HBaseServerExceptionPauseManager {
+  private static final Logger LOG = LoggerFactory.getLogger(HBaseServerExceptionPauseManager.class);
+
+  private final long pauseNs;
+  private final long pauseNsForServerOverloaded;
+  private final long timeoutNs;
+  private final long startNs;
+
+  public HBaseServerExceptionPauseManager(long pauseNs, long pauseNsForServerOverloaded,
+    long timeoutNs, long startNs) {
+    this.pauseNs = pauseNs;
+    this.pauseNsForServerOverloaded = pauseNsForServerOverloaded;
+    this.timeoutNs = timeoutNs;
+    this.startNs = startNs;
+  }
+
+  /**
+   * Returns the nanos, if any, for which the client should wait
+   * @param error The exception from the server
+   * @param tries The current retry count
+   * @return The time, in nanos, to pause. If empty then pausing would exceed our timeout, so we
+   *         should throw now
+   */
+  public OptionalLong getPauseNsFromException(Throwable error, int tries) {
+    long expectedSleepNs;
+    if (error instanceof RpcThrottlingException) {
+      RpcThrottlingException rpcThrottlingException = (RpcThrottlingException) error;
+      expectedSleepNs = TimeUnit.MILLISECONDS.toNanos(rpcThrottlingException.getWaitInterval());
+      long remainingTimeNs = remainingTimeNs();
+      if (expectedSleepNs > remainingTimeNs && remainingTimeNs > 0) {
+        if (LOG.isDebugEnabled()) {
+          LOG.debug("RpcThrottlingException suggested pause of {}ns which would exceed "
+            + "the timeout. We should throw instead.", expectedSleepNs, rpcThrottlingException);
+        }
+        return OptionalLong.empty();
+      }
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("Sleeping for {}ns after catching RpcThrottlingException", expectedSleepNs,
+          rpcThrottlingException);
+      }
+    } else {
+      expectedSleepNs =
+        HBaseServerException.isServerOverloaded(error) ? pauseNsForServerOverloaded : pauseNs;
+      // RpcThrottlingException tells us exactly how long the client should wait for,
+      // so we should not factor in the retry count for said exception
+      expectedSleepNs = getPauseTime(expectedSleepNs, tries - 1);
+    }
+
+    if (timeoutNs > 0) {
+      long remainingTimeNs = remainingTimeNs();
+      if (remainingTimeNs <= 0) {
+        return OptionalLong.empty();
+      }
+      expectedSleepNs = Math.min(remainingTimeNs, expectedSleepNs);
+    }
+
+    return OptionalLong.of(expectedSleepNs);
+  }
+
+  private long remainingTimeNs() {
```

Comment:
Do you think we can delete the `remainingTimeNs()` methods in the various caller classes now? Or are they still necessary? We could make this public if necessary and have any callers of the other to-be-deleted methods call this instead

Reply:
Ah actually, I'm realizing this gets a little bit awkward too because the scanner's startNs obviously isn't final. So we either need to support mutability of that field in the pause manager, reconstruct the pause manager on each call, or continue to leave the timeout responsibility outside of the pause manager. Do you have a preference?

Depending on the answer here we can certainly remove some of these remainingTimeNs methods

Reply:
I suppose you could pass in startNs as an argument, and only finalize timeoutNs in the constructor

Reply:
Regarding the redundancy of remainingTimeNs methods in the client, there's also a little bit of nuance in what these represent. Often at the client level the remainingTimeNs is just the difference between the timeout and the elapsed time, but in the context of the pause manager we also have to subtract the `SLEEP_DELTA_NS` (presumably to account for the lack of precision in sleeping for a number of millis). So just subbing one for the other isn't exactly 1:1

Reply:
Correct me if I'm wrong, but in the context of AsyncScanSingleRegionRpcRetryingCaller, `remainingTimeNs()` is only called once prior to this patch. The call moved into our new PauseManager code, so I assume that method is now unused.

In the context of AsyncBatchRpcRetryingCaller, it is called 3 times... so not unused at this point, but maybe could still be unified. I feel like we could do this:

1. Remove `startNs` arg from PauseManager constructor, since as you said it's not final for scans.
2. expose `PauseManager.remainingTimeNs()` as public or package protected, and add a `long startNs` argument.
3. Also add `long startNs` as argument in `getPauseNsFromException()` since it seems we need that for scans
4. In getPauseNsFromException(), use `remainingTimeNs(startNs) - SLEEP_DELTA_NS` (so the SLEEP_DELTA_NS gets removed from `PauseManager.remainingTimeNs(long startNs)` method
5. Replace all the calls to existing `remainingTimeNs()` methods in AsyncBatchCaller with `pauseManager.remainingTimeNs(startNs)`

Thoughts?

Reply:
totally onboard, just pushed https://github.com/apache/hbase/pull/5275/commits/902f8db18008d87901d4a96d654eb1de3a4991f0